specs/capella/beacon-chain.md

```diff
@@ -371,15 +272,29 @@ def process_block(state: BeaconState, block: BeaconBlock) -> None:
 
 ```python
 def process_withdrawals(state: BeaconState, payload: ExecutionPayload) -> None:
-    num_withdrawals = min(MAX_WITHDRAWALS_PER_PAYLOAD, len(state.withdrawal_queue))
-    dequeued_withdrawals = state.withdrawal_queue[:num_withdrawals]
-
-    assert len(dequeued_withdrawals) == len(payload.withdrawals)
-    for dequeued_withdrawal, withdrawal in zip(dequeued_withdrawals, payload.withdrawals):
-        assert dequeued_withdrawal == withdrawal
-
-    # Remove dequeued withdrawals from state
-    state.withdrawal_queue = state.withdrawal_queue[num_withdrawals:]
+    epoch = get_current_epoch(state)
+    index = state.last_withdrawal_validator_index
+    for withdrawal in payload.withdrawals:
+        index = ValidatorIndex((index + 1) % len(state.validators))
+        while index != withdrawal.validator_index:
```

Comment:
it seems like this is reintroducing a linear scan of the validator set, which I thought we wanted to remove?

Reply:
This linear scan succeeds in very few iterations unless the whole network is leaking for over 5 days. Clients are free to cache this however they want. It's very different than being forced to loop through the whole set